src/qobj/operators.jl

```diff
@@ -139,40 +140,181 @@ julia> fock(20, 4)' * a_d * fock(20, 3)
+    N = 2 * j
+    ((floor(N) != (N)) || (j < 0)) &&
+        throw(ArgumentError("The spin quantum number (j) j must be a non-negative integer or half-integer."))
+
+    σ = _jm(j)
+    return QuantumObject((σ' - σ) / 2im, Operator, [Int(N + 1)])
+end
+function jmat(j::Real, ::Val{:z})
+    N = 2 * j
+    ((floor(N) != (N)) || (j < 0)) &&
+        throw(ArgumentError("The spin quantum number (j) j must be a non-negative integer or half-integer."))
+
+    return QuantumObject(_jz(j), Operator, [Int(N + 1)])
+end
+function jmat(j::Real, ::Val{:+})
+    N = 2 * j
+    ((floor(N) != (N)) || (j < 0)) &&
+        throw(ArgumentError("The spin quantum number (j) j must be a non-negative integer or half-integer."))
+
+    return QuantumObject(adjoint(_jm(j)), Operator, [Int(N + 1)])
+end
+function jmat(j::Real, ::Val{:-})
+    N = 2 * j
+    ((floor(N) != (N)) || (j < 0)) &&
+        throw(ArgumentError("The spin quantum number (j) j must be a non-negative integer or half-integer."))
+
+    return QuantumObject(_jm(j), Operator, [Int(N + 1)])
+end
+jmat(j::Real, ::Val{T}) where {T} = throw(ArgumentError("Invalid spin operator: $(T)"))
+
+function _jm(j::Real)
+    m = j:(-1):-j
+    data = sqrt.(j * (j + 1) .- m .* (m .- 1))[1:end-1]
+    return spdiagm(-1 => Array{ComplexF64}(data))
+end
+_jz(j::Real) = spdiagm(0 => ComplexF64[j - k for k in 0:Int(2 * j)])
+
+@doc raw"""
+    spin_Jx(j::Real)
+
+``S_x`` operator for Spin-`j`, where `j` is the spin quantum number and can be a non-negative integer or half-integer
+
+See also [`jmat`](@ref).
+"""
+spin_Jx(j::Real) = jmat(j, Val(:x))
+
+@doc raw"""
+    spin_Jy(j::Real)
+
+``S_y`` operator for Spin-`j`, where `j` is the spin quantum number and can be a non-negative integer or half-integer
+
+See also [`jmat`](@ref).
+"""
+spin_Jy(j::Real) = jmat(j, Val(:y))
+
+@doc raw"""
+    spin_Jz(j::Real)
+
+``S_z`` operator for Spin-`j`, where `j` is the spin quantum number and can be a non-negative integer or half-integer
+
+See also [`jmat`](@ref).
+"""
+spin_Jz(j::Real) = jmat(j, Val(:z))
+
+@doc raw"""
+    spin_Jm(j::Real)
+
+``S_-`` operator for Spin-`j`, where `j` is the spin quantum number and can be a non-negative integer or half-integer
+
+See also [`jmat`](@ref).
+"""
+spin_Jm(j::Real) = jmat(j, Val(:-))
+
+@doc raw"""
+    spin_Jp(j::Real)
+
+``S_+`` operator for Spin-`j`, where `j` is the spin quantum number and can be a non-negative integer or half-integer
+
+See also [`jmat`](@ref).
+"""
+spin_Jp(j::Real) = jmat(j, Val(:+))
+
+@doc raw"""
+    spin_J_set(j::Real)
+
+A set of Spin-`j` operators ``(S_x, S_y, S_z)``, where `j` is the spin quantum number and can be a non-negative integer or half-integer
+
+Note that this functions is same as `jmat(j)`. See also [`jmat`](@ref).
+"""
+spin_J_set(j::Real) = jmat(j)
+
 @doc raw"""
     sigmap()
 
 Pauli ladder operator ``\hat{\sigma}_+ = \hat{\sigma}_x + i \hat{\sigma}_y``.
+
+See also [`jmat`](@ref).
 """
-sigmap() = destroy(2)
+sigmap() = jmat(0.5, Val(:+))
```

Comment:
Is it more efficient that the previous definition?

Reply:
I think for `sigmap` and `sigmam`, are not so different compared to the original methods.

However, I do think `sigmax`, `sigmay`, and `sigmaz` would be slightly more efficient.

Also, I personally think it is a bit weird to see `sigmap() = destroy(2)` lol.